compiler/qsc_eval/src/backend.rs

```diff
@@ -116,38 +130,86 @@ impl SparseSim {
     pub fn new() -> Self {
         Self {
             sim: QuantumSim::new(None),
+            noise: PauliNoise::default(),
+            rng: None,
+        }
+    }
+
+    #[must_use]
+    pub fn new_with_noise(noise: &PauliNoise) -> Self {
+        Self {
+            sim: QuantumSim::new(None),
+            noise: *noise,
+            rng: if noise.is_noiseless() {
+                None
+            } else {
+                Some(StdRng::from_entropy())
+            },
         }
     }
+
+    #[must_use]
+    fn is_noiseless(&self) -> bool {
+        self.rng.is_none()
+    }
+
+    fn apply_noise(&mut self, q: usize) {
+        if let Some(rng) = &mut self.rng {
+            let p = rng.gen_range(0.0..1.0);
+            if p >= self.noise.distribution[2] {
+                // In the most common case we don't apply noise
+            } else if p < self.noise.distribution[0] {
+                self.sim.x(q);
+            } else if p < self.noise.distribution[1] {
+                self.sim.y(q);
+            } else {
+                self.sim.z(q);
+            }
+        }
+        // No noise applied if rng is None.
+    }
 }
 
 impl Backend for SparseSim {
     type ResultType = bool;
 
     fn ccx(&mut self, ctl0: usize, ctl1: usize, q: usize) {
         self.sim.mcx(&[ctl0, ctl1], q);
+        self.apply_noise(ctl0);
+        self.apply_noise(ctl1);
+        self.apply_noise(q);
     }
 
     fn cx(&mut self, ctl: usize, q: usize) {
         self.sim.mcx(&[ctl], q);
+        self.apply_noise(ctl);
+        self.apply_noise(q);
     }
 
     fn cy(&mut self, ctl: usize, q: usize) {
         self.sim.mcy(&[ctl], q);
+        self.apply_noise(ctl);
+        self.apply_noise(q);
     }
 
     fn cz(&mut self, ctl: usize, q: usize) {
         self.sim.mcz(&[ctl], q);
+        self.apply_noise(ctl);
+        self.apply_noise(q);
     }
 
     fn h(&mut self, q: usize) {
         self.sim.h(q);
+        self.apply_noise(q);
     }
 
     fn m(&mut self, q: usize) -> Self::ResultType {
+        self.apply_noise(q);
         self.sim.measure(q)
     }
 
     fn mresetz(&mut self, q: usize) -> Self::ResultType {
+        self.apply_noise(q);
         let res = self.sim.measure(q);
         if res {
             self.sim.x(q);
```

Comment:
I hadn't thought of this before, but should `mresetz` (and by extension `reset`) include applying noise on the conditional bit flip? Do want returning to the ground state itself to be subject to probabilistic failure?

Reply:
I think it is, and my experimental branch had that. Now I think that allocation of fresh qubits should also be subject to this noise. I changed code to apply noise right after allocation and also after reset.

Reply:
After playing with it for a while I changed my mind. I will now remove noise on freshly allocated qubits and after reset. We will expose apply_noise to qshap. The user will be able to apply noise explicitly if this is needed.